presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java

```diff
@@ -1846,7 +1849,7 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         checkArgument(!handle.getAnalyzePartitionValues().isPresent() || constraint.getSummary().isAll(), "Analyze should not have a constraint");
 
         HivePartitionResult partitionResult = partitionManager.getPartitions(metastore, new HiveIdentity(session), handle, constraint);
-        HiveTableHandle newHandle = partitionManager.applyPartitionResult(handle, partitionResult);
+        HiveTableHandle newHandle = partitionManager.applyPartitionResult(handle, partitionResult, constraint.getColumns());
```

Comment:
`applyFilter` may be called multiple times throughout the optimization process, so this needs to combine the existing columns references in the handle, if any, with the new ones.

You should be able to reproduce this with a test that does something like:

```
SELECT * FROM (
    SELECT * FROM table
    WHERE partition_column = 'xxx'
)
WHERE non_partition_column = 'yyy'
```

Reply:
Good catch. I added a test like this which was failing. I found that I didn't need to combine the existing column references with the new ones as the LayoutConstraintEvaluator appears to already combine the expressions so we already get the combined set of columns from the constraint. Instead, I had to add a condition to the if statement below to compare handle.getConstraintColumns() and newHandle.getConstraintColumns(). Otherwise, you lose the constraint column information when Optional.empty() is returned. .

Reply:
There's no guarantee the predicates will all be applied in one shot. Actually, the case I was thinking of is more complicated to reproduce. It can happen when there's a join and each side is progressively narrowed down as the predicate pushdown logic is applied by predicates that are inferred from each side on each iteration. E.g. the left side produces a list of partitions that constrains the right side, which produces tighter predicate that further constrains the left side.